Advanced Tutorial.md

```diff
@@ -0,0 +1,211 @@
+
+MSON can be added to resources, actions, and individual requests or responses via an `+ Attributes` section. The random number generator API from above could look like this with MSON:
+
+```apib
+## Random Numbers [/random]
+
+### Generate a random number [GET]
+Generates and returns a random number.
+
++ Response 200 (application/json)
+
+    + Attributes
+
+        + value: 12345 (number)
+```
+
+When the above blueprint is parsed it will have a JSON body and JSON Schema example generated for it from the MSON attributes.
+
+The polls API question resource can also be modeled using MSON. By default, the `Attributes` section is assumed to be an object, but in this case we explicitly list it as an `array` instead:
+
+```apib
+### List All Questions [GET]
++ Response 200 (application/json)
+
+    + Attributes (array)
+
+        + (object)
+            + question: Favourite programming language? (required)
+            + published_at: `2014-11-11T08:40:51.620Z` (required)
+            + url: '/questions/1' (required)
+            + choices (array, required)
+                + (object)
+                    + choice: 'Javascript' (required)
+                    + url: '/questions/1/choices/1' (required)
+                    + votes: 2048 (number, required)
+```
+
+Note: MSON is still considered a beta feature!
+
+## Data Structures
+
+Once you start using MSON, you may find yourself wanting to reuse certain commonly used or nested data structure components. This is possible with the `## Data Structures` section. Attributes sections can then reference the data structures defined in the Data Structures section by name.
+
+Using the polls API question resource, we can split out the `Question` and `Choice` objects:
+
+```apib
+### List All Questions [GET]
++ Response 200 (application/json)
+
+    + Attributes (array[Question])
+
+## Data Structures
+
+### Question
++ question: Favourite programming language? (required)
++ published_at: `2014-11-11T08:40:51.620Z` (required)
++ url: '/questions/1' (required)
++ choices (array[Choice], required)
+
+### Choice
++ choice: 'Javascript' (required)
++ url: '/questions/1/choices/1' (required)
++ votes: 2048 (number, required)
+```
+
+## Relation Types
+
+Actions in a blueprint can define a semantic domain-specific meaning by defining a relation type using a `+ Relation` section. This means that an action can have a specific meaning regardless of its URI or name, and allows clients to be built based on the domain of the API rather than specific URIs.
+
+For example, in the polls API:
+
+```apib
+## Question [/question/{id}]
+### View a Question Detail [GET]
++ Relation: question
+
+### Update a Question [PUT]
++ Relation: question_update
```

Comment:
The Polls API doesn't allow you to update a question and I don't think it makes sense given our polls domain. Questions shouldn't be modifiable since users can vote on them.

I'd suggest we drop this action from the example.